qubes/ext/core_features.py

```diff
@@ -104,7 +104,90 @@ async def qubes_features_request(self, vm, event, untrusted_features):
             if _version_re.fullmatch(untrusted_value):
                 vm.features["qubes-agent-version"] = untrusted_value
+
+        # handle boot mode advertisement
+        old_bootmode_info = {}
+        for feature_key, feature_val in vm.features.items():
+            if feature_key.startswith(
+                "boot-mode.kernelopts."
+            ) or feature_key.startswith("boot-mode.name"):
+                old_bootmode_info[feature_key] = feature_val
+        new_bootmode_info = {}
+        untrusted_sanitized_features = {}
+        for (
+            untrusted_feature_key,
+            untrusted_feature_value,
+        ) in untrusted_features.items():
+            if not all(c in string.printable for c in untrusted_feature_value):
+                continue
+            untrusted_sanitized_features[untrusted_feature_key] = (
+                untrusted_feature_value
+            )
+        for (
+            untrusted_feature_key,
+            untrusted_feature_value,
+        ) in untrusted_sanitized_features.items():
+            if untrusted_feature_key.startswith("boot-mode.kernelopts."):
+                bootmode_name = untrusted_feature_key.split(".")[2]
+                if bootmode_name == "":
+                    continue
+                bootmode_feature = untrusted_feature_key
+                bootmode_value = untrusted_feature_value
+                new_bootmode_info[bootmode_feature] = bootmode_value
+        for (
+            untrusted_feature_key,
+            untrusted_feature_value,
+        ) in untrusted_sanitized_features.items():
+            if untrusted_feature_key.startswith("boot-mode.name."):
+                bootmode_name = untrusted_feature_key.split(".")[2]
+                if bootmode_name == "":
+                    continue
+                if (
+                    f"boot-mode.kernelopts.{bootmode_name}"
+                    not in new_bootmode_info
+                ):
+                    continue
+                bootmode_feature = untrusted_feature_key
+                bootmode_value = untrusted_feature_value
+                new_bootmode_info[bootmode_feature] = bootmode_value
+        # Only allow a maximum of 64 boot modes, fail to add or remove any
+        # boot modes if more than 64 would be defined in the end
+        if len(new_bootmode_info) <= 64:
+            # Prevent wiping all boot modes
+            if new_bootmode_info:
+                for feature_key in old_bootmode_info:
+                    if feature_key not in new_bootmode_info:
+                        del vm.features[feature_key]
+                for feature_key, feature_val in new_bootmode_info.items():
+                    vm.features[feature_key] = feature_val
+        for (
+            untrusted_feature_key,
+            untrusted_feature_value,
+        ) in untrusted_sanitized_features.items():
+            if untrusted_feature_key == "boot-mode.active":
```

Comment:
There is no need to iterate over all just for those two keys. Simply do `if "boot-mode.active" in untrusted_features` or similar.